src/metabase/models/data_permissions.clj

```diff
@@ -631,6 +631,38 @@
                           :group_id group-id))
             lowest-to-highest-values))))
 
+(defenterprise new-group-view-data-permission-level
+  "Returns the default view-data permission level for a new group for a given database. On OSS, this is always `unrestricted`."
+  metabase-enterprise.advanced-permissions.common
+  [_group-id]
+  :unrestricted)
+
+(defn set-new-group-permissions!
+  "Sets permissions for a newly-added group to their appropriate values for a single database. This is generally based
+  on the permissions of the All Users group."
+  [group-or-id db-or-id all-users-group-id]
+  (let [db-id                (u/the-id db-or-id)
+        view-data-level      (new-group-view-data-permission-level db-id)
+        create-queries-level (or (->> (t2/select-fn-set :value
+                                                        [:model/DataPermissions [:perm_value :value]]
+                                                        :perm_type :perms/create-queries
+                                                        :db_id db-id
+                                                        :group_id all-users-group-id)
+                                      (coalesce-most-restrictive :perms/create-queries))
+                                 :query-builder-and-native)
+        download-level      (or (->> (t2/select-fn-set :value
+                                                       [:model/DataPermissions [:perm_value :value]]
+                                                       :perm_type :perms/download-results
+                                                       :db_id db-id
+                                                       :group_id all-users-group-id)
+                                     (coalesce-most-restrictive :perms/download-results))
+                                :one-million-rows)]
+    (set-database-permission! group-or-id db-or-id :perms/view-data view-data-level)
+    (set-database-permission! group-or-id db-or-id :perms/create-queries create-queries-level)
+    (set-database-permission! group-or-id db-or-id :perms/download-results download-level)
+    (set-database-permission! group-or-id db-or-id :perms/manage-table-metadata :no)
+    (set-database-permission! group-or-id db-or-id :perms/manage-database :no)))
```

Comment:
Take this or leave it, but I can’t help to think this could be more data oriented. It makes things easier to repl when a function returns data. I know our testing framework lets us emulate that by putting stuff into the database, and taking it out as a map, but it does add some friction when trying to figure out why a perm got a certain value.

Concretely, it's easy to imagine this split into 2 functions, one that **computes** the permissions (can test/see all combos) and one that **inserts them** (basically a doseq over that map that uses `set-database-permission!`.

Reply:
I pulled out the perm-computing logic into a separate `new-group-permissions` function 👍